Unwrap *some* t.co links on twitter

cowlicks opened the pull request:
When this is more stable, I'll move the content scipt out of `first_parties.js` and into its own file. And have the content script only inject on twitter.

ghostwords commented on src/manifest.json:
```diff
@@ -44,6 +44,16 @@
+        "js/first_parties.js"
+      ],
+      "matches": [
+        "<all_urls>"
```
Chrome injected scripts can access page DOM fine (they do have a [separate JS environment](https://developer.chrome.com/extensions/content_scripts#execution-environment), but we don't need that for our purposes here). So we don't need to make use of the "injected script injecting a page script" pattern.

ghostwords replied:
I suggest injecting the Twitter-specific script directly, and only for Twitter URLs (international too). This way we avoid making any changes whatsoever to any non-Twitter pages.

cowlicks replied:
:+1:

ghostwords commented on src/js/first_parties/twitter.js:
```diff
@@ -0,0 +1,37 @@
+      <span class="tco-ellipsis"><span class="invisible">&nbsp;</span>…</span>
+    </a>
+  
+*/
```
Should remove the comment to save a little RAM on Twitter pages.

cowlicks replied:
👍

ghostwords commented on src/js/first_parties/twitter.js:
```diff
@@ -0,0 +1,37 @@
+*/
+
+function unwrapTwitterURLs() {
+  // will this still select link with different capitalization like "HhtPs://t.Co"? 
```
Twitter probably won't be inserting malformed `t.co` links so we don't have to worry about this, but should be easy to test (modify link manually in Dev Tools' Inspector, run replacer code manually) ...

cowlicks replied:
For posterity, the answer was no. But you can add an `i` in your query to make it case insensitive https://drafts.csswg.org/selectors-4/#attribute-case

cowlicks commented:
Comments addressed.

Some other things:
* I couldn't find a list of international twitter domains. An example like "twitter.co.uk" redirects to "twitter.com" for me, but that might be due to my location. Does twitter actually use different TLDs? 
* Profiled this and it has a very low impact ~%0.6 of cpu usage.
* twitter has an event handler for all clicks on the page, the minified js is hard to follow. But I don't see any extra network requests triggered by clicking on a link. I'll add an click eventlistener to elements we modify override theirs.

cowlicks commented:
I added a listener to the links to stop the propagation of the click event up the DOM to twitter.

ghostwords commented on src/tests/tests/first_parties.js:
```diff
@@ -0,0 +1,32 @@
+    let el = document.createElement("div");
+    el.setAttribute("href", tco);
+    el.setAttribute("data-expanded-url", destination);
+    document.querySelectorAll = function () {
```
We never restore the original, so it stays overwritten for all subsequent tests. Also, might as well use `sinon.stub` to overwrite and restore, since we already use Sinon.

ghostwords replied:
We should overwrite and restore stuff via QUnit module `beforeEach` and `afterEach` properties, to avoid dealing with async test headaches.

ghostwords commented on src/tests/tests/first_parties.js:
```diff
@@ -0,0 +1,32 @@
+(function() {
+  QUnit.test("test twitter content script", (assert) => {
+    const NUM_CHECKS = 1;
+    let done = assert.async(NUM_CHECKS);
```
Nit: `assert.async()` defaults to 1, NUM_CHECKS is unnecessary here.

ghostwords commented on src/tests/tests/first_parties.js:
```diff
@@ -0,0 +1,32 @@
+    let script = document.createElement("script");
+    script.src = "../js/first_parties/twitter.js";
+    script.onload = function () {
+      unwrapTwitterURLs();
```
Hmm, `setInterval` doesn't call right away, right? But we do want to call right away, for testing here and production use. So let's change our code and get rid of this line.

ghostwords replied:
Also, our linter is catching `unwrapTwitterURLs` here as being undefined, which short-circuits and fails all automated tests.

In this case we have a false positive that you can tell ESLint to ignore via the globals directive at the top of the file (`/* globals unwrapTwitterURLs:false */`), but this line should get removed so it won't be necessary.

You should get set up for local linting so that you can catch these errors/warnings early. Your best bet should be to set up your editor to lint as you save (or even as you type).

ghostwords commented on src/js/first_parties/twitter.js:
```diff
@@ -0,0 +1,16 @@
+  }
+}
+
+setInterval(unwrapTwitterURLs, 5000);
```
We probably want to call `unwrapTwitterURLs` right away, right? But `setInterval` delays first execution too. Also, generally speaking (maybe not so much here), `setInterval` is risky since it's not guaranteed to execute on time or evenly spaced or anything so it can lead to your callbacks piling up. Safer to repeat by having the function call itself with `setTimeout` at the end.

ghostwords commented on src/js/first_parties/twitter.js:
```diff
@@ -0,0 +1,16 @@
+    let deu = elem.getAttribute("data-expanded-url");
+    if (deu && (deu.startsWith("https://") || deu.startsWith("http://"))) {
+      elem.href = deu;
+      elem.addEventListener("click", function (e) {
```
Have you tested that this successfully prevents page listeners from receiving the click? Wondering if this won't work because of the content script's separate JS execution context.

cowlicks replied:
I tested that it works, yeah. I put a break point on the click action and verified we no longer passed control to twitter. I was trying to unittest it, but was having trouble making the mocked event listeners get called in the right order.

ghostwords commented on src/tests/tests/first_parties.js:
```diff
@@ -0,0 +1,32 @@
+
+    let qf = document.getElementById("qunit-fixture");
+    qf.appendChild(script);
+    qf.remove();
```
Unnecessary to remove, I think.

ghostwords commented on src/tests/tests/first_parties.js:
```diff
@@ -0,0 +1,32 @@
+(function() {
```
No module?

cowlicks replied:
No, because this is a content script, it has to run when it is loaded.

cowlicks replied:
Ooops, you meant the Qunit module. I'll add one.

ghostwords replied:
Sorry, I meant QUnit module, to help us organize our tests and also so that we can use stuff like 'beforeEach'.

ghostwords commented on src/js/first_parties/twitter.js:
```diff
@@ -0,0 +1,16 @@
+    if (deu && (deu.startsWith("https://") || deu.startsWith("http://"))) {
+      elem.href = deu;
+      elem.addEventListener("click", function (e) {
+        e.stopPropagation(); // don't let the event click propogate to twitter
```
Typo ("propagate"), but I dunno if need any comments here.

ghostwords commented on src/tests/tests/first_parties.js:
```diff
@@ -0,0 +1,32 @@
+    let tco = "http://t.co/beach-detour/";
+    let destination = "https://the.beach/";
+    let el = document.createElement("div");
+    el.setAttribute("href", tco);
```
I don't think this works ... `el.href = tco;` does though.

ghostwords commented:
@pde We should probably strip referrer headers here too, right? @cowlicks suggested "rel=noreferrer" might be enough.

ghostwords commented:
Here are a few existing `t.co` link unwrappers/deobfuscators:
- https://github.com/Spaxe/Goodbye--t.co-
- https://github.com/dlo/no-t.co
- https://github.com/tony-o/t.co-bypass ([HN thread](https://news.ycombinator.com/item?id=8071618))

ghostwords commented:
Let's test with [TweetDeck](https://en.wikipedia.org/wiki/TweetDeck) too. We may need to use another link property as fallback, as in the extensions linked above.

cowlicks commented:
I added, the `noreferrer` stuff and a test for it, as well as tweetdeck support.

cowlicks commented:
Tested manually that noreferrer, unwrapping, and blocking click tracking work on chrome on twitter/tweeted. Tested noreferrer and unwrapping in firefox too, but wasn't sure how to test click tracking there.

ghostwords commented:
We should handle links in "cards" (or whatever those things are called). Seems like a common scenario. For example: https://twitter.com/NILC_org/status/867763812845240325.

ghostwords commented:
Also, `t.co` can be http apparently: https://twitter.com/xkcdrss/status/255158017253838848

ghostwords commented on src/js/first_parties/twitter.js:
```diff
@@ -31,7 +31,7 @@ function unwrapTwitterURLs() {
 }
 
-if (typeof wasrun === "undefined" || !wasrun) {
+if (typeof wasrun === "undefined" || !wasrun) {  // eslint-disable-line no-undef
```
You can avoid having to use an ESLint directive by checking `window.has_run`.

cowlicks replied:
👍

ghostwords commented:
Looks like tests are failing because the `src/js/first_parties.js` wasn't included in `src/tests/index.html` and it's causing a "global failure" in QUnit after merging with master. It's actually a JavaScript error (in the console of  tests/index.html) in this branch too, just not getting caught by QUnit for whatever reason.

ghostwords commented:
If I'm scrolling down a Twitter feed and quickly mousing about, I can still see `t.co` URLs ... I'm not sure how annoying this will be to users. Might be OK, but it is noticeable.

ghostwords commented:
Not sure about adding an entire source file that adds another listener and runs `extractHostFromURL` on every page URL just to support TweetDeck without running multiple/more complex DOM queries (or duplicating content script code). Doesn't feel good to roll our own content script injector.

cowlicks commented:
@ghostwords So one good reason to keep using `chrome.tabs.executeScript` is that it lets us parameterize the content script. This is nice because otherwise, they are really hard to test (notice we don't have any unittests for our other content scripts). So I can remove `chrome.tabs.executeScript` but I'll also probably just have to remove the unittests that I added.

ghostwords commented:
Sorry, I don't understand why testing requires `executeScript`/needing to set vars.

Separately, I am not sure unit tests are a good fit for verifying `t.co` replacement keeps working. Unit tests might catch when we mess something up in our content script functions, but will not catch the more likely scenario, Twitter changing something on their end. Seems like Selenium testing would be a better fit to verify integrated functionality.

cowlicks commented:
@ghostwords How should we handle authenticating on twitter?

cowlicks commented:
@ghostwords I've removed the injector code and the tests. If have a suggestions about how to implement functional tests for this please let me know. Currently I'm wondering:

* what exactly should I test (we can't do a simple "are all the t.co's gone?")
* how should we authenticate with twitter, or what account should we use

ghostwords commented:
What do you mean by "authenticating on twitter"? What are we trying to do and why do we need to authenticate?

The test could be to load a Twitter page (that we control so that the tweet content is stable) and verify that the links that we expect to be there (we can [find them by link text](http://selenium-python.readthedocs.io/locating-elements.html#locating-hyperlinks-by-link-text) maybe) are indeed there, and that the links point to expected full URLs, not `t.co` URLs.

ghostwords commented:
What were your findings working with `MutationObserver`? User experience-wise (scrolling Twitter and clicking on links) does our solution compare favorably to existing projects I linked above?

ghostwords commented on src/js/firstparties/twitter.js:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  removeInDoc(document);
+  let iframes = document.getElementsByTagName('iframe');
```
Why do this ourselves instead of letting Chrome do it with `all_frames: true` in manifest.json?

cowlicks replied:
Because then those content scripts won't share the global `found_tcos_query` and `fixes` variables that they need to replace links which were found in other iframes.

ghostwords replied:
And the reason we need to share that state is because sometimes (in "cards"?) the link is inside the iframe but the property containing the full URL is outside. This is also why the replacement is a two step process in the code.

ghostwords commented on src/js/firstparties/twitter.js:
```diff
@@ -0,0 +1,62 @@
+  maybeAddNoreferrer(tco);
+}
+
+function checkLink(linkElelement) {
```
Typo: `linkElelement`

cowlicks commented:
@ghostwords I added back the tests. Fixed typo. Please let me know if there is anything else.

To recap our offline discussion about `MutationObserver`. To make it performant, we'd have to tailor specifically to twitter's DOM, making it more fragile to changes in twitter.

ghostwords commented on src/tests/tests/firstparties.js:
```diff
@@ -0,0 +1,71 @@
+    let tweet = makeTweet(attribute);
+
+    // set the config for twitter.js
+    window.config = {'queryParam': attribute};
```
We can remove this, right?

ghostwords commented on src/tests/tests/firstparties.js:
```diff
@@ -0,0 +1,71 @@
+  let tco = 'http://t.co/beach-detour/';
+  let destination = 'https://the.beach/';
+
+  document.querySelectorAllBefore = document.querySelectorAll;
```
We should probably make whatever hacks we need to make in `before` or `beforeEach` properties of the [QUnit module](https://api.qunitjs.com/QUnit/module). This way the stubbing doesn't happen for all tests, just the ones it should happen for.

ghostwords replied:
Should we also stub `setTimeout` to prevent the content script's timer from running continuously on the tests page?

cowlicks replied:
I added a stub for `setInterval`. I don't think the stubs should be added to `before`/`beforeEach` since they are specific to one test (not every test, or the module).

ghostwords commented:
Looks like we have a memory leak ... scroll through a Twitter feed while watching memory use in Chrome's Task Manager. Twitter memory use goes up without Privacy Badger but seems to go up faster with Privacy Badger. I think the reason is we never clean up `fixes` or `found_tcos_query` so they just keeps getting bigger and bigger. Long `found_tcos_query` strings also seem to impact CPU.

ghostwords commented:
We should double-check memory/CPU use (from scrolling a Twitter feed for a while) on Firefox specifically, as you usually feel CPU issues more on Firefox.

ghostwords commented:
I am not sure the JS click listener prevention bit actually works on Twitter. It does on our tests page, but maybe only because the script is running in the same context as the page vs. where running in the extension content script context on Twitter. For example, run the following snippet on Twitter and then scroll through a feed to get some fresh links:
```javascript
function links() {
  let els = document.querySelectorAll("a[href^='https://t.co/']");
  for (let i = 0; i < els.length; i++) {
    els[i].addEventListener('click', function() {
      alert('ok');
    })
  }
}
links();
setInterval(links, 500);
```

ghostwords commented:
I think we can ship without JS listener prevention, just raising this now so we don't make any claims we can't back up.

cowlicks commented:
@ghostwords the listener I add only stops the click from being propagated up the dom. Event listener's on the element itself still here the click. Twitters' click listener is on the body tag. This is why in the test I put the click listener on the `fixture` instead of the `tweet`.

cowlicks commented:
@ghostwords ready for review

ghostwords commented:
Will you also make the unit test tweaks?

ghostwords commented:
I just pushed a couple of updates to the [unwrap-tco branch](https://github.com/EFForg/privacybadger/compare/unwrap-tco). We weren't picking up (homepage) links in the profile sidebar (they have a title attribute), or `pic.twitter.com` links (we can use their link text contents). Could you pull into your branch and review my update please?

Looking good otherwise! Will test it some more and review performance next.

cowlicks commented:
@ghostwords looking at the added code for unwrapping links in the profile url: since we are just trying to unwrap a link in one particular case, could we better just add a `unwrapProfileUrl` function that would just use a specific css selector to find it? Instead of adding to the to the main `unwrapTwitterURLs` function. I expect that adding to more parts to the `tcos_with_destination_query` would add non-trivial cpu cycles. And having a more specific function would make us less likely to break twitter if something about the `title` attribute changes. Could you add a `unwrapProfileUrl` function in a seperate PR?

Another concern is that using `title` might open up some XSS issues, since `title` is a standard link property while `data-url-expanded` is an attribute. Users generally have more control over what gets parsed into a `title` field than what goes in `data-expanded-url`.

For the `pic.twitter.com` stuff, I don't think those are necessary, since they already point to twitter. And also I don't have as much trust in users not being able to manipulate `textContent`.

cowlicks commented:
This seems fine on Firefox to me. Anything else?

ghostwords commented:
Good feedback, let's punt on my updates for now. For the record, here is the commit in question: https://github.com/EFForg/privacybadger/commit/f626a6cc7d9f84f91beee1b38e9009628138c17b.
```diff
diff --git a/src/js/firstparties/twitter.js b/src/js/firstparties/twitter.js
index c01ff80c..68307457 100644
--- a/src/js/firstparties/twitter.js
+++ b/src/js/firstparties/twitter.js
@@ -1,5 +1,11 @@
-let query_param = 'data-expanded-url';
-let tcos_with_destination = "a[" + query_param + "][href^='https://t.co/'], a[" + query_param + "][href^='http://t.co/']";
+let query_params = [
+  'data-expanded-url',
+  'title',
+];
+let tcos_with_destination = query_params.map(val => {
+  return "a[" + val + "][href^='https://t.co/'], a[" + val + "][href^='http://t.co/']";
+}).join(", ");
+
 let fixes = {};
 
 function maybeAddNoreferrer(link) {
@@ -19,6 +25,15 @@ function unwrapTco(tco, destination) {
   maybeAddNoreferrer(tco);
 }
 
+function getOriginalUrl(link) {
+  for (let i = 0; i < query_params.length; i++) {
+    let attr = link.getAttribute(query_params[i]);
+    if (attr && (attr.startsWith("https://") || attr.startsWith("http://"))) {
+      return attr;
+    }
+  }
+}
+
 function findInAllFrames(query) {
   let out = [];
   document.querySelectorAll(query).forEach((node) => {
@@ -38,15 +53,17 @@ function findInAllFrames(query) {
 
 function unwrapTwitterURLs() {
   findInAllFrames(tcos_with_destination).forEach((link) => {
-    let attr = link.getAttribute(query_param);
-    if (attr && (attr.startsWith("https://") || attr.startsWith("http://"))) {
-      fixes[link.href] = attr;
-      unwrapTco(link, attr);
+    let orig_url = getOriginalUrl(link);
+    if (orig_url) {
+      fixes[link.href] = orig_url;
+      unwrapTco(link, orig_url);
     }
   });
   findInAllFrames("a[href^='https://t.co/'], a[href^='http://t.co/'").forEach((link) => {
     if (fixes.hasOwnProperty(link.href)) {
       unwrapTco(link, fixes[link.href]);
+    } else if (link.textContent.startsWith("pic.twitter.com/")) {
+      unwrapTco(link, 'https://' + link.textContent);
     }
   });
 }
```
